storage/redis_adapter.go

```diff
@@ -3,13 +3,67 @@ package storage
 import (
 	"fmt"
 	"log"
+	"time"
 
 	r "gopkg.in/redis.v3"
 )
 
+type message struct {
+	app         string
+	messageBody string
+}
+
+func newMessage(app string, messageBody string) *message {
+	return &message{
+		app:         app,
+		messageBody: messageBody,
+	}
+}
+
+type messagePipeliner struct {
+	bufferSize    int
+	messageCount  int
+	pipeline      *r.Pipeline
+	timeoutTicker *time.Ticker
+	queuedApps    map[string]bool
+}
+
+func newMessagePipeliner(bufferSize int, redisClient *r.Client) *messagePipeliner {
+	return &messagePipeliner{
+		bufferSize:    bufferSize,
+		pipeline:      redisClient.Pipeline(),
+		timeoutTicker: time.NewTicker(time.Second),
+		queuedApps:    map[string]bool{},
+	}
+}
+
+func (mp *messagePipeliner) addMessage(message *message) {
+	if err := mp.pipeline.RPush(message.app, message.messageBody).Err(); err == nil {
+		mp.queuedApps[message.app] = true
+		mp.messageCount++
+	} else {
+		log.Printf("Error adding rpush to %s to the pipeline: %s", message.app, err)
```

Comment:
Ordinarily, I would, but there can be many handlers (each its own goroutine) calling the storage adapter's `Write()` function, and in turn calling this function. The error can only be returned so far up the call stack before there isn't anyone waiting for a possible error. Logging it right as it occurs seemed the easiest way to mitigate that. The harder way would be introducing an error channel, I suppose. Thoughts?